hedera-node/test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenAssociationSpecs.java

```diff
@@ -119,6 +129,59 @@ public boolean canRunConcurrent() {
         return true;
     }
 
+    @HapiTest
+    public HapiSpec canHandleInvalidAssociateTransactions() {
+        final String alice = "ALICE";
+        final String bob = "BOB";
+        final String unknownID = "0.0." + Long.MAX_VALUE;
+        return defaultHapiSpec("CanHandleInvalidAssociateTransactions")
+                .given(
+                        newKeyNamed(MULTI_KEY),
+                        cryptoCreate(alice),
+                        cryptoCreate(bob),
+                        cryptoDelete(bob),
+                        tokenCreate(VANILLA_TOKEN),
+                        tokenCreate(KNOWABLE_TOKEN),
+                        tokenAssociate(alice, KNOWABLE_TOKEN),
+                        tokenCreate(TBD_TOKEN).adminKey(MULTI_KEY),
+                        tokenDelete(TBD_TOKEN))
+                .when()
+                .then(
+                        tokenAssociate(null, VANILLA_TOKEN)
+                                .fee(DEFAULT_FEE)
+                                .signedBy(DEFAULT_PAYER)
+                                .hasPrecheck(INVALID_ACCOUNT_ID),
+                        tokenAssociate(unknownID, VANILLA_TOKEN)
+                                .fee(DEFAULT_FEE)
+                                .signedBy(DEFAULT_PAYER)
+                                .hasPrecheck(INVALID_ACCOUNT_ID),
+                        tokenAssociate(bob, VANILLA_TOKEN).fee(DEFAULT_FEE).hasKnownStatus(ACCOUNT_DELETED),
+                        tokenAssociate(alice, List.of()).hasKnownStatus(SUCCESS),
+                        tokenAssociate(alice, VANILLA_TOKEN, VANILLA_TOKEN)
+                                .hasPrecheck(TOKEN_ID_REPEATED_IN_TOKEN_LIST),
+                        tokenAssociate(alice, unknownID).hasKnownStatus(INVALID_TOKEN_ID),
+                        tokenAssociate(alice, TBD_TOKEN).hasKnownStatus(TOKEN_WAS_DELETED),
+                        tokenAssociate(alice, KNOWABLE_TOKEN).hasKnownStatus(TOKEN_ALREADY_ASSOCIATED_TO_ACCOUNT));
```

Comment:
These permutations feel like they belong in unit tests. But, if they do, then why didn't the unit tests we have catch this? 🤷 

I'm nervous that adding a large number of negative-case hapi tests for some 50 transaction handlers is going to severely increase the amount of time these tests take. But I'm not sure what to do about it...

Reply:
It is a good question why no unit tests caught these errors. I believe part of the answer is that we redefined the purpose of `pureChecks()` and `preHandle()` multiple times while the handlers were already in development.

Having these tests is valuable, in my opinion, because we will move the checks around to do as little as possible during ingest. These tests will ensure that, overall, the errors are still handled as expected. Also, by introducing these tests, we fix which error code is returned in a given scenario, and we can document it.